builder/processor-tools/src/main/java/io/helidon/builder/processor/tools/BuilderTypeTools.java

```diff
@@ -113,26 +108,63 @@ public Optional<TypeInfo> createTypeInfo(
-                                                                                         processingEnv.getElementUtils()))
+                                   .annotations(
+                                           createAnnotationAndValueListFromElement(element, processingEnv.getElementUtils()))
                                    .elementInfo(elementInfo)
                                    .otherElementInfo(otherElementInfo)
+                                   .referencedTypeNamesToAnnotations(
+                                           toReferencedTypeNamesAndAnnotations(
+                                                   processingEnv, typeName, elementInfo, otherElementInfo))
                                    .modifierNames(modifierNames)
                                    .update(it -> toTypeInfo(annotationTypeName, element, processingEnv, wantDefaultMethods)
                                            .ifPresent(it::superTypeInfo))
                                    .build());
     }
 
+    @SuppressWarnings("unchecked")
+    private Map<TypeName, List<AnnotationAndValue>> toReferencedTypeNamesAndAnnotations(ProcessingEnvironment processingEnv,
+                                                                                        TypeName typeName,
+                                                                                        Collection<TypedElementName>... refs) {
+        Map<TypeName, List<AnnotationAndValue>> result = new LinkedHashMap<>();
+        for (Collection<TypedElementName> ref : refs) {
+            for (TypedElementName typedElementName : ref) {
+                collectReferencedTypeNames(result, processingEnv, typeName, List.of(typedElementName.typeName()));
+                collectReferencedTypeNames(result, processingEnv, typeName, typedElementName.typeName().typeArguments());
+            }
+        }
+        return result;
+    }
+
+    private void collectReferencedTypeNames(Map<TypeName, List<AnnotationAndValue>> result,
+                                            ProcessingEnvironment processingEnv,
+                                            TypeName typeName,
+                                            Collection<TypeName> referencedColl) {
+        for (TypeName referenced : referencedColl) {
+            if (isBuiltInJavaType(referenced) || typeName.equals(referenced)) {
+                continue;
+            }
+
+            result.compute(referenced, (k, v) -> {
+                if (v == null) {
+                    // first time processing, we only need to do this on pass #1
+                    TypeElement typeElement = processingEnv.getElementUtils().getTypeElement(k.name());
+                    if (typeElement != null) {
+                        v = createAnnotationAndValueListFromElement(typeElement, processingEnv.getElementUtils());
+                    }
+                }
+                return v;
+            });
```

Comment:
Since you want to process that only once, Would not `computeIfAbsent` better suite you needs?

Reply:
next push.